File: kotlin/src/main/kotlin/benchmarks/CsvParse.kt
package benchmarks

import Benchmark
import Helper
import com.opencsv.CSVReader
import java.io.StringReader
import java.util.Locale

class CsvParse : Benchmark() {
    private lateinit var data: String
    private var resultVal: UInt = 0u
    private val rows = configInt("rows")

    override fun name(): String = "CSV::Parse"

    override fun prepare() {
        val sb = StringBuilder(rows * 50)

        for (i in 0 until rows) {
            val c = 'A' + i % 26
            val x = Helper.nextFloat()
            val z = Helper.nextFloat()
            val y = Helper.nextFloat()

            sb
                .append('"')
                .append("point ")
                .append(c)
                .append("\\n, \"\"")
                .append(i % 100)
                .append("\"\"\"")
                .append(',')

            sb.append("%.10f".format(Locale.US, x)).append(',')

            sb.append(',')

            sb.append("%.10f".format(Locale.US, z)).append(',')

            sb
                .append('"')
                .append('[')
                .append(if (i % 2 == 0) "true" else "false")
                .append("\\n, ")
                .append(i % 100)
                .append(']')
                .append('"')
                .append(',')

            sb.append("%.10f".format(Locale.US, y)).append('\n')
        }

        data = sb.toString()
    }

    data class Point(
        val x: Double,
        val y: Double,
        val z: Double,
    )

    private fun parsePoints(csvData: String): List<Point> {
        val points = mutableListOf<Point>()

        CSVReader(StringReader(csvData)).use { reader ->
            var record: Array<String>? = reader.readNext()
            while (record != null) {
                val x = record[1].toDouble()
                val z = record[3].toDouble()
                val y = record[5].toDouble()
                points.add(Point(x, y, z))
                record = reader.readNext()
            }
        }

        return points
    }

    override fun run(iterationId: Int) {
        val points = parsePoints(data)

        if (points.isEmpty()) return

        var xSum = 0.0
        var ySum = 0.0
        var zSum = 0.0

        for (p in points) {
            xSum += p.x
            ySum += p.y
            zSum += p.z
        }

        val count = points.size.toDouble()
        val xAvg = xSum / count
        val yAvg = ySum / count
        val zAvg = zSum / count

        resultVal += Helper.checksumF64(xAvg) + Helper.checksumF64(yAvg) + Helper.checksumF64(zAvg)
    }

    override fun checksum(): UInt = resultVal
}

File: rust/src/benchmarks/csv_parse.rs
use super::super::{helper, Benchmark};
use crate::config_i64;
use simd_csv;
use std::fmt::Write;

#[derive(Debug, Clone, Copy)]
struct Point {
    x: f64,
    y: f64,
    z: f64,
}

pub struct CsvParse {
    rows: usize,
    checksum: u32,
    data: String,
}

impl CsvParse {
    pub fn new() -> Self {
        let rows = config_i64("CSV::Parse", "rows") as usize;
        Self {
            rows,
            checksum: 0,
            data: String::new(),
        }
    }

    fn generate_data(&mut self) {
        self.data.clear();

        for i in 0..self.rows {
            let c = (b'A' + (i % 26) as u8) as char;
            let x = helper::next_float(1.0);
            let z = helper::next_float(1.0);
            let y = helper::next_float(1.0);
            write!(&mut self.data, "\"point {c}\\n, \"\"{}\"\"\",", i % 100).unwrap();
            write!(&mut self.data, "{:.10},", x).unwrap();
            write!(&mut self.data, ",").unwrap();
            write!(&mut self.data, "{:.10},", z).unwrap();
            let flag = if i % 2 == 0 { "true" } else { "false" };
            write!(&mut self.data, "\"[{flag}\\n, {}]\",", i % 100).unwrap();
            write!(&mut self.data, "{:.10}\n", y).unwrap();
        }
    }

    fn parse_points(&self) -> Vec<Point> {
        let mut points = Vec::new();

        let mut reader = simd_csv::ZeroCopyReaderBuilder::new()
            .has_headers(false)
            .delimiter(b',')
            .quote(b'"')
            .from_reader(self.data.as_bytes());

        while let Ok(Some(record)) = reader.read_byte_record() {
            let x = std::str::from_utf8(&record[1])
                .unwrap()
                .parse::<f64>()
                .unwrap();
            let z = std::str::from_utf8(&record[3])
                .unwrap()
                .parse::<f64>()
                .unwrap();
            let y = std::str::from_utf8(&record[5])
                .unwrap()
                .parse::<f64>()
                .unwrap();

            points.push(Point { x, y, z });
        }

        points
    }
}

impl Benchmark for CsvParse {
    fn name(&self) -> String {
        "CSV::Parse".to_string()
    }

    fn prepare(&mut self) {
        self.generate_data();
    }

    fn run(&mut self, _iteration_id: i64) {
        let points = self.parse_points();

        if points.is_empty() {
            return;
        }

        let (mut x_sum, mut y_sum, mut z_sum) = (0.0, 0.0, 0.0);
        for point in &points {
            x_sum += point.x;
            y_sum += point.y;
            z_sum += point.z;
        }

        let count = points.len() as f64;
        let x_avg = x_sum / count;
        let y_avg = y_sum / count;
        let z_avg = z_sum / count;

        self.checksum = self
            .checksum
            .wrapping_add(helper::checksum_f64(x_avg))
            .wrapping_add(helper::checksum_f64(y_avg))
            .wrapping_add(helper::checksum_f64(z_avg));
    }

    fn checksum(&self) -> u32 {
        self.checksum
    }
}
